File: examples/c/app-token-ring/app-token-ring.c
/* Copyright (c) 2008-2020. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

#include "simgrid/actor.h"
#include "simgrid/engine.h"
#include "simgrid/host.h"
#include "simgrid/mailbox.h"

#include "xbt/log.h"
#include "xbt/str.h"
#include "xbt/sysdep.h"

#include <stddef.h>
#include <stdio.h>

XBT_LOG_NEW_DEFAULT_CATEGORY(app_token_ring, "Messages specific for this msg example");

/* Main function of all actors used in this example */
static void relay_runner(int argc, XBT_ATTRIB_UNUSED char* argv[])
{
  xbt_assert(argc == 0, "The relay_runner function does not accept any parameter from the XML deployment file");

  const char* name = sg_actor_self_get_name();
  int rank         = xbt_str_parse_int(name, "Any actor of this example must have a numerical name, not %s");

  sg_mailbox_t my_mailbox = sg_mailbox_by_name(name);

  /* The last actor sends the token back to rank 0, the others send to their right neighbor (rank+1) */
  char neighbor_mailbox_name[256];
  snprintf(neighbor_mailbox_name, 255, "%d", rank + 1 == sg_host_count() ? 0 : rank + 1);

  sg_mailbox_t neighbor_mailbox = sg_mailbox_by_name(neighbor_mailbox_name);

  char* res;
  if (rank == 0) {
    /* The root actor (rank 0) first sends the token then waits to receive it back */
    XBT_INFO("Host \"%d\" send 'Token' to Host \"%s\"", rank, neighbor_mailbox_name);
    sg_mailbox_put(neighbor_mailbox, xbt_strdup("Token"), 1000000);

    res = (char*)sg_mailbox_get(my_mailbox);
    XBT_INFO("Host \"%d\" received \"%s\"", rank, res);
  } else {
    /* The others actors receive from their left neighbor (rank-1) and send to their right neighbor (rank+1) */
    res = (char*)sg_mailbox_get(my_mailbox);
    XBT_INFO("Host \"%d\" received \"%s\"", rank, res);
    XBT_INFO("Host \"%d\" send 'Token' to Host \"%s\"", rank, neighbor_mailbox_name);
    sg_mailbox_put(neighbor_mailbox, xbt_strdup("Token"), 1000000);
  }
  free(res);
}

int main(int argc, char* argv[])
{
  simgrid_init(&argc, argv);
  xbt_assert(argc > 1, "Usage: %s platform.xml\n", argv[0]);
  simgrid_load_platform(argv[1]); /* - Load the platform description */

  size_t host_count = sg_host_count();
  sg_host_t* hosts  = sg_host_list();

  XBT_INFO("Number of hosts '%zu'", host_count);
  for (size_t i = 0; i < host_count; i++) {
    /* - Give a unique rank to each host and create a @ref relay_runner process on each */
    char* name_host  = bprintf("%zu", i);
    sg_actor_t actor = sg_actor_init(name_host, hosts[i]);
    sg_actor_start(actor, relay_runner, 0, NULL);
    free(name_host);
  }
  free(hosts);

  simgrid_run();

  XBT_INFO("Simulation time %g", simgrid_get_clock());
  return 0;
}

File: examples/java/app_token_ring/app_token_ring.java
/* Copyright (c) 2007-2025. The SimGrid Team. All rights reserved.          */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

import org.simgrid.s4u.*;

class RelayRunner extends Actor {
  public void run() throws SimgridException
  {
    long token_size = 1000000; /* The token is 1MB long */
    Mailbox my_mailbox;
    Mailbox neighbor_mailbox;
    int rank;

    try {
      rank = Integer.parseInt(this.get_name());
    } catch (NumberFormatException e) {
      throw new NumberFormatException("Actors of this example must have a numerical name, not " + this.get_name());
    }
    my_mailbox = this.get_engine().mailbox_by_name(String.valueOf(rank));
    if (rank + 1 == this.get_engine().get_host_count())
      /* The last actor sends the token back to rank 0 */
      neighbor_mailbox = this.get_engine().mailbox_by_name("0");
    else
      /* The others actors send to their right neighbor (rank+1) */
      neighbor_mailbox = this.get_engine().mailbox_by_name(String.valueOf(rank + 1));

    if (rank == 0) {
      /* The root actor (rank 0) first sends the token then waits to receive it back */
      Engine.info("Host \"%d\" send 'Token' to Host \"%s\"", rank, neighbor_mailbox.get_name());
      String msg = "Token";
      neighbor_mailbox.put(msg, token_size);
      String res = (String)my_mailbox.get();
      Engine.info("Host \"%d\" received \"%s\"", rank, res);
    } else {
      String res = (String)my_mailbox.get();
      Engine.info("Host \"%d\" received \"%s\"", rank, res);
      Engine.info("Host \"%d\" send 'Token' to Host \"%s\"", rank, neighbor_mailbox.get_name());
      neighbor_mailbox.put(res, token_size);
    }
  }
}

public class app_token_ring {
  public static void main(String[] args)
  {
    Engine e = new Engine(args);
    if (args.length < 1)
      Engine.die("Usage: %s platform.xml\n");
    e.load_platform(args[0]);

    Engine.info("Number of hosts '%d'", e.get_host_count());
    int id = 0;
    for (Host host : e.get_all_hosts()) {
      /* - Give a unique rank to each host and create a RelayRunner actor on each */
      host.add_actor(String.valueOf(id), new RelayRunner());
      id++;
    }
    e.run();
    Engine.info("Simulation time %g", Engine.get_clock());

    // The following call is useless in your code, but our continuous integration uses it to track memleaks
    e.force_garbage_collection();
  }
}
